### PR TITLE
[TEST] Fix elasticsearch-security-cli JUnit runner

### PR DESCRIPTION
**Pull request by justincr-elastic**
Using the JUnit test runner in this project failed when launched in
an IDE. A "Jar Hell" error message was logged to console.

Root cause was BC non-FIPS and BC FIPS on the test classpath at the
same time. BC non-FIPS jars are imported by the project as api
dependencies, and BC-FIPS jars were inherited as testImplementation
dependencies from x-pack core module.

JUnit test runner failures were triggered from ESTestCase. That super
class loads BootstrapForTesting, which detects and rejects BC non-FIPS
vs BC-FIPS on the classpath. Those two sets of jars conflict because
they contain duplicate class names with different implementations.

To fix the JUnit test runner error, build.gradle has been updated to
exclude the BC-FIPS testImplementation dependencies inherited
from x-pack core module. After the fix, the JUnit test runner works
in an IDE for the elasticsearch-security-cli project.

Reason for the fix is the JUnit test runner may be more convenient,
or faster, to use in an IDE compared to the Gradle test runner.

See also: #69403

### CODE REVIEW

**Comment by justincr-elastic** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by justincr-elastic** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/cli/build.gradle
Comment (jkakavas) [inline x-pack/plugin/security/cli/build.gradle:22]: we don't have bcpg-fips in `core` right ? So this exclusion is not needed.

**Review by jkakavas — APPROVED** [review]
LGTM pending the one comment below. I was wondering if we should declare the dependency differently in https://github.com/elastic/elasticsearch/blob/35ec6f348c396cec742c987716c30d79f973c2e8/x-pack/plugin/core/build.gradle#L53 so that it doesn't leak into `testArtifact`. Out of curiosity, I will defer to @elastic/es-delivery for an opinion, but I'm not sure it matters that much. The suggested solutions here works fine!

File: x-pack/plugin/security/cli/build.gradle
Comment (justincr-elastic) [reply x-pack/plugin/security/cli/build.gradle:22]: bcpg-fips is part of plugin-cli with bc-fips.
```
[root@d0747cd87f3f plugin-cli]# ls -l
total 4000
-r--r--r-- 1 elasticsearch root 3775682 Oct  8  2020 bc-fips-1.0.2.jar
-r--r--r-- 1 elasticsearch root  271757 Oct  8  2020 bcpg-fips-1.0.4.jar
-r--r--r-- 1 elasticsearch root   43784 Aug 26 09:02 elasticsearch-plugin-cli-7.14.1.jar
```

elasticsearch-security-cli does not inherit bcpg-fips from core, so it is more precautionary to exclude both bcpg-fips and bc-fips in elasticsearch-security-cli.

Comment (jkakavas) [reply x-pack/plugin/security/cli/build.gradle:22]: > bcpg-fips is part of plugin-cli with bc-fips.

Yeap, this is defined in `distribution/tools/plugin-cli/build.gradle`

```
  api "org.bouncycastle:bcpg-fips:1.0.4"
  api "org.bouncycastle:bc-fips:1.0.2"
```

I guess my point is that since `testArtifact(project(xpackModule('core')))` doesn't include `bcpg-fips`, there is need to exclude it here. As you pointed out, `bcpg-fips` is a dependency only for `plugin-clie`. We are pretty declarative and explicit in our dependency tracking, I think there is no need to proactively exclude a dependency in an unrelated package.

**Comment by mark-vieira** [comment]
My other question is how do those tests not explode via JarHell when we run via `-Dtests.fips.enabled=true` since that would similarly put both jars on the classpath. I think the root issue is that when we run with FIPS we hackily smash those jars onto the classpath rather than modeling them properly. That's the whole reason we need this `compileOnly` dependency in the first case.

Also another great reason why we should be using test fixtures, and not these "test artifact" dependencies because we end up pulling in way more than we need to :frowning: 

I'll defer to @breskeby on what the best coarse of action is here. Removing the `compileOnly` dependencies from the exported test artifact won't always work. In this case, we don't expose any of the classes via the API but we might in other cases, in which case compiling the consuming project would fail. Perhaps that's ok? Rene, what's the default behavior for `compileOnly` dependencies when consuming from another project? They come through, right, but are simply not also included on the runtime classpath?

**Comment by tvernum** [comment]
> My other question is how do those tests not explode via JarHell 

The answer will not delight you...

https://github.com/elastic/elasticsearch/blob/v7.14.1/x-pack/plugin/security/cli/build.gradle#L38-L53

**Comment by tvernum** [comment]
I'm going to drop from the review list here. If @jkakavas and Delivery are happy, you don't need my opinions as well.

**Comment by jkakavas** [comment]
> > My other question is how do those tests not explode via JarHell
> 
> The answer will not delight you...
> 
> https://github.com/elastic/elasticsearch/blob/v7.14.1/x-pack/plugin/security/cli/build.gradle#L38-L53

Not delighted but not surprised either :D https://github.com/elastic/elasticsearch/pull/51733

**Review by ywangd — APPROVED** [review]
LGTM

Thanks for getting to the bottom of this! I am OK with this approach since it fixes the problem at hand. Alternatively, I wonder whether we could eliminate the bc-fips jars from `core`'s testArtifact when the test is not `inFipsJvm` (something like conditional dependency and compliation in `core` module). It has the advantage that individual consumer does not need to worry about excluding the jars on its own. So we don't get into this problem in another place or future.

**Comment by breskeby** [comment]
> My other question is how do those tests not explode via JarHell when we run via `-Dtests.fips.enabled=true` since that would similarly put both jars on the classpath. I think the root issue is that when we run with FIPS we hackily smash those jars onto the classpath rather than modeling them properly. That's the whole reason we need this `compileOnly` dependency in the first case.
> 
> Also another great reason why we should be using test fixtures, and not these "test artifact" dependencies because we end up pulling in way more than we need to 😦
> 
> I'll defer to @breskeby on what the best coarse of action is here. Removing the `compileOnly` dependencies from the exported test artifact won't always work. In this case, we don't expose any of the classes via the API but we might in other cases, in which case compiling the consuming project would fail. Perhaps that's ok? Rene, what's the default behavior for `compileOnly` dependencies when consuming from another project? They come through, right, but are simply not also included on the runtime classpath?

the gradle default is that only `api` dependencies are exposed to the compile classpath of a consuming project. The runtime classpath contains `api` and `implementation` but not `compileOnly`

File: x-pack/plugin/security/cli/build.gradle
Comment (breskeby) [reply x-pack/plugin/security/cli/build.gradle:22]: I agree with @jkakavas that the exclude should not be required. How does this actually solve the problem with the security cli junit runner?

**Comment by breskeby** [comment]
I think the way to go forward to a proper fix is to only expose the api dependencies of the testartifact. That allows proper encapsulation of compile and runtime dependencies and is actually what api dependencies are about. You should not depend on a compileOnly or implementation dependency at compile time. 
I'll look into a proper solution next week after the long weekend as the week is ending here already. 
If this PR solves a big pain point for you NOW I don't want to block you on not merging here. I'll create a follow up PR with a tweaked testArtifact handling in the upcoming days

File: x-pack/plugin/security/cli/build.gradle
Comment (justincr-elastic) [reply x-pack/plugin/security/cli/build.gradle:22]: Thank you for the clarification that we should be explicit about exclusions. I will remove the pcpg-fips exclusion.

**Comment by justincr-elastic** [comment]
@elasticmachine update branch

**Comment by mark-vieira** [comment]
> The answer will not delight you...
>
> https://github.com/elastic/elasticsearch/blob/v7.14.1/x-pack/plugin/security/cli/build.gradle#L38-L53

At least I'm not crazy :wink:

**Comment by mark-vieira** [comment]
> I think the way to go forward to a proper fix is to only expose the api dependencies of the testartifact. That allows proper encapsulation of compile and runtime dependencies and is actually what api dependencies are about. You should not depend on a compileOnly or implementation dependency at compile time.

Yep, this makes sense to me. Essentially, `compileOnly` is an `implementation` dependency that's also _not_ inherited by `runtimeClasspath`. :+1:

**Comment by breskeby** [comment]
@justincr-elastic @jkakavas I think this PR is outdated as the original problem should have been fixed by now via https://github.com/elastic/elasticsearch/pull/78238 can you confirm this is no longer an issue for you? I just tested it manually in my idea setup with master

**Comment by jkakavas** [comment]
@justincr-elastic can we close this ?

**Comment by justincr-elastic** [comment]
Closing because it is addressed by the other PR.

**Pull request closed without merging**
